File: 117. Populating Next Right Pointers in Each Node II.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, left, right, next):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        curr = dummy = Node(0, None, None, None)
        head = root
        
        while root:
            if root.left:
                curr.next = root.left
                curr = curr.next
            
            if root.right:
                curr.next = root.right
                curr = curr.next
            
            root = root.next
            
            if not root:
                curr = dummy
                root = dummy.next
                dummy.next = None
        
        return head

File: 117. Populating Next Right Pointers in Each Node II.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}

    public Node(int _val,Node _left,Node _right,Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/
class Solution {
    public Node connect(Node root) {
        Node dummy, curr, head;
        dummy = new Node(0, null, null, null);
        curr=dummy; head=root;
        while (head!=null) {
            if (head.left!=null) {
                curr.next = head.left;
                curr = curr.next;
            }
            if (head.right!=null) {
                curr.next = head.right;
                curr = curr.next;
            }
            head = head.next;
            if (head == null) {
                head = dummy.next;
                curr = dummy;
                dummy.next = null;
            }
        }
        return root;
    }
}